lib/vnet/customdnszonevalidator.go

```diff
@@ -0,0 +1,71 @@
+package vnet
+
+import (
+	"context"
+	"log/slog"
+	"net"
+	"slices"
+	"sync"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	clusterTXTRecordPrefix = "teleport-cluster="
+)
+
+type lookupTXTFunc = func(ctx context.Context, domain string) (txtRecords []string, err error)
+
+// customDNSZoneValidator validates custom DNS zones by making sure that they have a DNS TXT record of
+// "teleport-cluster=<cluster-name>" for the cluster in which they are used. This is meant to avoid the
+// possibility of a rogue application advertising a public_addr with a DNS name not controlled by the cluster
+// admins, which could be used to trick VNet users.
+//
+// After finding that a zone is valid once, this is cached indefinitely. Invalid zones are a misconfiguration
+// so we don't cache negative results
+type customDNSZoneValidator struct {
+	lookupTXT  lookupTXTFunc
+	validZones map[string]struct{}
+	mu         sync.RWMutex
+}
+
+func newCustomDNSZoneValidator(lookupTXT lookupTXTFunc) *customDNSZoneValidator {
+	if lookupTXT == nil {
+		var resolver net.Resolver
+		lookupTXT = resolver.LookupTXT
+	}
+	return &customDNSZoneValidator{
+		lookupTXT:  lookupTXT,
+		validZones: make(map[string]struct{}),
+	}
+}
+
+// validate returns an error if [customDNSZone] is not valid for [clusterName].
+func (c *customDNSZoneValidator) validate(ctx context.Context, clusterName, customDNSZone string) error {
+	c.mu.RLock()
+	_, ok := c.validZones[customDNSZone]
+	c.mu.RUnlock()
+	if ok {
+		return nil
+	}
+
+	requiredTXTRecord := clusterTXTRecordPrefix + clusterName
+	slog.InfoContext(ctx, "Checking validity of custom DNS zone by querying for required TXT record.", "zone", customDNSZone, "record", requiredTXTRecord)
+
+	records, err := c.lookupTXT(ctx, customDNSZone)
+	if err != nil {
+		return trace.Wrap(err, "looking up TXT records for %q", customDNSZone)
```

Comment:
Another thing: when I was just messing around with Let's Encrypt setup, I added a CNAME for my custom DNS zone. Later I wanted to add a TXT record. That's when I realized that you cannot have a CNAME and pretty much any other record set at the same time.

Do you think it's going to cause problems? Looking at DNS zones for some of the domains that I own, I see that for example Mandrill asks you to set up a TXT record under `mandrill._domainkey.<your domain>` instead of just `<your domain>`.